### PR TITLE
Log warning if SIP is disabled and CLI version is < 2.15.1

### PR DESCRIPTION
**Pull request by angelapwen**
The `macos-latest` image is now ARM rather than Intel, so we need to change some of our PR Checks. The build tracer on CLI versions before v2.15.1 did not support the ARM machines where System Integrity Protection was disabled, which now includes `macos-latest`.

This change:
- uses `macos-12` Intel runners for any checks where the CLI version is below v2.15.1
- logs a warning if SIP is disabled, and if the CLI version is below v2.15.1

Separately, the `macos-latest` image no longer supports Go on the path by default, so this PR adds `setup-go` to all PR checks analyzing Go. 

I've updated the `Required` PR checks on `main` to include the new ones, but have not updated `v2` or `v3` yet.

### Merge / deployment checklist

- [x] Confirm this change is backwards compatible with existing workflows.
- [x] Confirm the [readme](https://github.com/github/codeql-action/blob/main/README.md) has been updated if necessary.
- [x] Confirm the [changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md) has been updated if necessary.

### CODE REVIEW

File: src/init-action.ts [hunk 1]
```diff
@@ -467,6 +473,18 @@ async function run() {
...
 
+    // For CLI versions <2.15.1, build tracing caused errors in MacOS ARM machines with
+    // System Integrity Protection (SIP) disabled.
+    if (
+      !(await codeQlVersionAbove(codeql, "2.15.1")) &&
+      process.platform === "darwin" &&
+      !(await isSipEnabled(logger))
+    ) {
+      logger.warning(
+        "CodeQL versions 2.15.0 and lower are not supported on MacOS ARM machines with System Integrity Protection (SIP) disabled.",
```
Comment (adityasharad) [inline src/init-action.ts:484]: The warning mentions ARM, but you're not checking `process.arch`.
Either we can change the warning to say macOS in general with SIP disabled is not supported on <=2.15.0 (not strictly true, but I don't know if we fixed other relocation issues that would affect Intel),
or change the code above to check `process.arch` being `arm` or `arm64`.

File: src/init-action.ts [hunk 2]
```diff
@@ -467,6 +473,18 @@ async function run() {
       }
     }
 
+    // For CLI versions <2.15.1, build tracing caused errors in MacOS ARM machines with
+    // System Integrity Protection (SIP) disabled.
+    if (
+      !(await codeQlVersionAbove(codeql, "2.15.1")) &&
```
Comment (adityasharad) [inline src/init-action.ts:479]: Minor, separate: I had to go look at the definition to remind myself whether this was `>` or `>=`. Perhaps we should rename it `codeQlVersionAtLeast` or similar.

File: pr-checks/sync.py [hunk 3]
```diff
@@ -81,7 +89,8 @@ def writeHeader(checkStream):
             # Ensure that this is serialized as a folded (`>`) string to preserve the readability
             # of the generated workflow.
             'if': FoldedScalarString(textwrap.dedent('''
-                    matrix.os == 'macos-latest' && (
+                    (matrix.os == 'macos-latest' || 
+                    matrix.os == 'macos-12') && (
```
Comment (adityasharad) [inline pr-checks/sync.py:93]: Can we simplify this to just `runner.os == 'macOS'`?

**Review by adityasharad — COMMENTED** [review]
Thanks - generally looks good but a couple of minor suggestions.

File: src/init-action.ts (hunk 1 above)
Comment (angelapwen) [reply src/init-action.ts:484]: Ah yes, good point 👍 will change it to `process.arch` as I believe that's more accurate to the problem we were looking at in 2.15.1.

File: src/init-action.ts (hunk 2 above)
Comment (angelapwen) [reply src/init-action.ts]: Good point, I always double check that it's using `gte` too. I've made the change

File: pr-checks/sync.py (hunk 3 above)
Comment (angelapwen) [reply pr-checks/sync.py]: Yep, done!

File: src/init-action.ts [hunk 4]
```diff
@@ -477,7 +477,7 @@ async function run() {
     // System Integrity Protection (SIP) disabled.
     if (
       !(await codeQlVersionAbove(codeql, "2.15.1")) &&
-      process.platform === "darwin" &&
```
Comment (adityasharad) [inline src/init-action.ts]: This line needs to remain!

Comment (angelapwen) [reply src/init-action.ts]: Oh yes, Linux exists 😆

**Review by adityasharad — APPROVED** [review]

**Pull request merged**